pontoon/base/models.py

```diff
@@ -1675,6 +1687,87 @@ class ChangedEntityLocale(models.Model):
 
     class Meta:
         unique_together = ('entity', 'locale')
+        
+
+class EntityFiltersManager(models.Manager):
+    def update_filters_state(self, entity, locale):
+        """
+        Update state of filters of entity for a given locale.
+        """
+        filters_instance = self.get_queryset().get(entity=entity)
+
+        forms_count = locale.nplurals if entity.string_plural else 1
+        translations = entity.translation_set.filter(locale=locale).values_list('string', 'approved', 'fuzzy')
+
+        def set_field(field_name, value):
+            setattr(filters_instance, '{}_{}'.format(locale.code, field_name), value)
+
+        set_field('status', self.get_entity_status(translations, locale))
+        set_field('unchanged', self.is_unchanged(entity, translations, forms_count))
+        set_field('has_suggestions', self.has_suggestions(translations))
+        filters_instance.save()
+
+    def get_entity_status(self, translations, forms_count):
+        """
+        Calculate current status of an entity.
+        """
+
+        if forms_count == len([1 for _, approved, _ in translations if approved]):
+            return 'translated'
+
+        elif forms_count == len([1 for _, _, fuzzy in translations if fuzzy]):
+            return 'fuzzy'
+
+        elif len([1 for _, approved, fuzzy in translations if not approved and not fuzzy]):
+            return 'suggested'
+
+        return 'missing'
+
+    def is_unchanged(self, entity, translations, forms_count):
+        """
+        Checks if entity has been changed.
+        """
+        strings = [s for s, _, _ in translations]
+        same_string = entity.string in strings
+        if same_string and forms_count > 1:
+            return entity.string_plural in strings
+        return same_string
+
+    def has_suggestions(self, translations):
+        return bool(len([1 for _, approved, fuzzy in translations if not approved and not fuzzy]))
+
+
+class EntityFilters(models.Model):
+    """
+    EntityFilters contains computed filters for every value.
+    Because of the performance reasons, every entity has one additional row in database
+    """
+    entity = models.ForeignKey(Entity)
+    status_choices = (
+        ('missing', 'missing'),
+        ('fuzzy', 'fuzzy'),
+        ('translated', 'translated'),
+        ('suggested', 'suggested')
+    )
+    extra_filters = (
+        'unchanged',
+        'has_suggestions'
+    )
+
+    objects = EntityFiltersManager()
+
+
+for locale_code in locales.CODES:
+    code = locale_code.lower().replace('-', '_')
```

Comment:
It's because this loop will add fields to EntityFilters and because of the compatibility with python properties don't allow underscores.